File: src/types/schemas.ts
// ─────────────────────────────────────────────────────────────
// Input Triggers
// ─────────────────────────────────────────────────────────────

/** A sudden, abnormal increase in network or request traffic. */
export interface TrafficSpike {
  /** ISO-8601 timestamp when the spike was detected. */
  timestamp: string;
  /** The source identifier (IP, subnet, service name, etc.). */
  source: string;
  /** Requests-per-second at the time of detection. */
  requests_per_second: number;
  /** Normal baseline RPS for this source. */
  baseline_rps: number;
  /** Severity score (0.0 – 1.0). */
  severity: number;
}

/** A burst of consecutive authentication failures. */
export interface FailedLogins {
  /** ISO-8601 timestamp of the most recent failure. */
  timestamp: string;
  /** The source identifier (IP, user-agent fingerprint, etc.). */
  source: string;
  /** How many failures inside the detection window. */
  attempt_count: number;
  /** Length of the detection window in seconds. */
  time_window_secs: number;
  /** Targeted account names / emails. */
  target_accounts: string[];
  /** Severity score (0.0 – 1.0). */
  severity: number;
}

/** Top-level anomaly report envelope sent to the Tauri backend. */
export interface AnomalyReport {
  /** Human-readable session label, e.g. "Session 1". */
  session_id: string;
  /** Discriminator: "traffic_spike" | "failed_logins". */
  trigger_type: "traffic_spike" | "failed_logins";
  /** JSON payload matching the trigger_type schema. */
  payload: TrafficSpike | FailedLogins;
}

// ─────────────────────────────────────────────────────────────
// Memory Artifact
// ─────────────────────────────────────────────────────────────

/** What HIRA remembers after resolving an incident. */
export interface MemoryArtifact {
  /** Unique incident identifier (UUID v4). */
  incident_id: string;
  /** The trigger type that spawned this incident. */
  trigger_type: string;
  /** Embedding vectors from the analysis pipeline. */
  vectors: number[];
  /** Whether the mitigation was successful. */
  mitigation_success: boolean;
  /** Free-text retrospective note. */
  hindsight_note: string;
  /** ISO-8601 timestamp of creation. */
  created_at: string;
}

// ─────────────────────────────────────────────────────────────
// CascadeFlow Audit Trail
// ─────────────────────────────────────────────────────────────

/** A single decision node in the CascadeFlow chain. */
export interface CascadeBlock {
  /** Step index within the cascade. */
  step: number;
  /** Which agent / module produced this block. */
  agent: string;
  /** Action taken: "route" | "mitigate" | "escalate" | etc. */
  action: string;
  /** Input data snapshot evaluated at this step. */
  input_snapshot: Record<string, unknown>;
  /** Decision / output produced by the agent. */
  output: Record<string, unknown>;
  /** Confidence score (0.0 – 1.0). */
  confidence: number;
  /** ISO-8601 timestamp. */
  timestamp: string;
}

/** Full audit trail for one incident's CascadeFlow. */
export interface CascadeFlowTrail {
  /** The incident this trail belongs to. */
  incident_id: string;
  /** Ordered decision blocks. */
  blocks: CascadeBlock[];
  /** Overall status. */
  status: "in_progress" | "completed" | "failed";
}

// ─────────────────────────────────────────────────────────────
// Baseline Analysis Result
// ─────────────────────────────────────────────────────────────

/** Returned by the Phase 1 "dumb" baseline command. */
export interface BaselineResult {
  session_id: string;
  trigger_type: string;
  recommendation: string;
  used_memory: boolean;
  used_cascade: boolean;
}

File: src-tauri/src/schemas.rs
use serde::{Deserialize, Serialize};

// ─────────────────────────────────────────────────────────────
// Input Triggers — the raw anomaly signals that enter HIRA
// ─────────────────────────────────────────────────────────────

/// A sudden, abnormal increase in network or request traffic.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrafficSpike {
    /// ISO-8601 timestamp when the spike was detected.
    pub timestamp: String,
    /// The source identifier (IP, subnet, service name, etc.).
    pub source: String,
    /// Requests-per-second at the time of detection.
    pub requests_per_second: f64,
    /// Normal baseline RPS for this source — used to gauge severity.
    pub baseline_rps: f64,
    /// Severity score derived upstream (0.0 – 1.0).
    pub severity: f64,
}

/// A burst of consecutive authentication failures.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FailedLogins {
    /// ISO-8601 timestamp of the most recent failure in the burst.
    pub timestamp: String,
    /// The source identifier (IP, user-agent fingerprint, etc.).
    pub source: String,
    /// How many failures occurred inside the detection window.
    pub attempt_count: u32,
    /// Length of the detection window in seconds.
    pub time_window_secs: u64,
    /// List of targeted account names / emails.
    pub target_accounts: Vec<String>,
    /// Severity score derived upstream (0.0 – 1.0).
    pub severity: f64,
}

/// The top-level anomaly report envelope sent from the frontend.
/// `trigger_type` discriminates the variant; `payload` carries
/// the JSON-serialised trigger body.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnomalyReport {
    /// A human-readable session label, e.g. "Session 1".
    pub session_id: String,
    /// Discriminator: "traffic_spike" | "failed_logins".
    pub trigger_type: String,
    /// The raw JSON payload matching the trigger_type schema.
    pub payload: serde_json::Value,
}

// ─────────────────────────────────────────────────────────────
// Memory Artifact — what HIRA remembers after resolving an
// incident (Phase 2+ will persist these; Phase 1 defines shape)
// ─────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryArtifact {
    /// Unique incident identifier (UUID v4).
    pub incident_id: String,
    /// The original trigger type that spawned this incident.
    pub trigger_type: String,
    /// Embedding vectors produced by the analysis pipeline.
    /// Stored as a flat f64 array; dimensionality is implicit.
    pub vectors: Vec<f64>,
    /// Whether the chosen mitigation was ultimately successful.
    pub mitigation_success: bool,
    /// Free-text retrospective note added after resolution.
    pub hindsight_note: String,
    /// ISO-8601 timestamp of artifact creation.
    pub created_at: String,
}

// ─────────────────────────────────────────────────────────────
// CascadeFlow Audit Trail — an ordered chain of decision blocks
// that records *why* HIRA chose a particular response path
// ─────────────────────────────────────────────────────────────

/// A single decision node in the CascadeFlow chain.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CascadeBlock {
    /// Monotonically increasing step index within the cascade.
    pub step: u32,
    /// Which agent / module produced this block.
    pub agent: String,
    /// What action was taken (e.g. "route", "mitigate", "escalate").
    pub action: String,
    /// The input data snapshot that was evaluated at this step.
    pub input_snapshot: serde_json::Value,
    /// The decision / output produced by the agent.
    pub output: serde_json::Value,
    /// Confidence score the agent assigned to its decision (0.0 – 1.0).
    pub confidence: f64,
    /// ISO-8601 timestamp of this step.
    pub timestamp: String,
}

/// The full audit trail for one incident's CascadeFlow.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CascadeFlowTrail {
    /// The incident this trail belongs to.
    pub incident_id: String,
    /// Ordered sequence of decision blocks.
    pub blocks: Vec<CascadeBlock>,
    /// Overall status: "in_progress" | "completed" | "failed".
    pub status: String,
}

// ─────────────────────────────────────────────────────────────
// Baseline Analysis Result — returned by the "dumb" Phase 1
// command before memory / CascadeFlow is wired up
// ─────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BaselineResult {
    pub session_id: String,
    pub trigger_type: String,
    pub recommendation: String,
    pub used_memory: bool,
    pub used_cascade: bool,
}
